[libc][malloc] Reuse the prev_ field for allocated blocks

mysterymath opened the pull request:
This applies a standard trick from Knuth for storing boundary tags with only one word of overhead for allocated blocks. The prev_ block is now only valid if the previous block is free.

This is safe, since only coalescing with a free node requires walking the blocks backwards. To allow determining whether it's safe to traverse backwards, the used flag is changed to a prev_free flag. Since it's still possible to unconditionally traverse forward, the prev_free flag for the next block can be used wherever the old used flag is, so long as there is always a next block.

To ensure there is always a next block, a sentinel last block is added at the end of the range of blocks. Due to the above, this costs only a single word per heap. This sentinel essentially just stores whether the last real block of the heap is free. The sentinel is always considered used and to have a zero inner size.

This completes the block optimizations needed to address #98086. The block structure should now be size-competitive with dlmalloc, although there are still a couple of broader fragmentation concerns to address.

mysterymath committed:
[libc][malloc] Reuse the prev_ field for allocated blocks

This applies a standard trick from Knuth for storing boundary tags with
only one word of overhead for allocated blocks. The prev_ block is now
only valid if the previous block is free.

This is safe, since only coalescing with a free node requires walking
the blocks backwards. To allow determining whether it's safe to traverse
backwards, the used flag is changed to a prev_free flag. Since it's
still possible to unconditionally traverse forward, the prev_free flag
for the next block can be used wherever the old used flag is, so long as
there is always a next block.

To ensure there is always a next block, a sentinel last block is added
at the end of the range of blocks. Due to the above, this costs only a
single word per heap. This sentinel essentially just stores whether the
last real block of the heap is free. The sentinel is always considered
used and to have a zero inner size.

This completes the block optimizations needed to address #98086. The
block structure should now be size-competitive with dlmalloc, although
there are still a couple of broader fragmentation concerns to address.

mysterymath committed:
clang-format

PiJoules commented on libc/src/__support/block.h:
```diff
@@ -182,40 +209,39 @@ class Block {
+    next()->next_ |= PREV_FREE_MASK;
+    // The next block's prev_ field becomes alive, as it is no longer part of
+    // this block's used space.
+    *new (&next()->prev_) offset_type = outer_size();
```
If the next block's `prev_` becomes alive, does this block's `inner_size()` need to also account for that? Asking since `inner_size()` looks to be unconditionally adding `sizeof(prev_)`, but perhaps that should be gated on also checking if `next()->next_` has the `PREV_FREE_MASK` set or not(?)

mysterymath replied:
The semantics of `inner_size()` were subtly recast to be the size available for allocation; the `prev_` field of the next block is always available, so it's always included in the inner size, regardless of whether or not the block is free. This allows checks against `inner_size()` to mostly persist unaltered, but it does complicate using it for block arithmetic.

PiJoules commented on libc/src/__support/block.h:
```diff
@@ -332,29 +353,34 @@ Block<OffsetType, kAlign>::init(ByteSpan region) {
 
   region = result.value();
-  if (region.size() < BLOCK_OVERHEAD)
+  if (region.size() < 2 * BLOCK_OVERHEAD)
```
Out of curiosity, why does this need to be `2 * BLOCK_OVERHEAD`?

mysterymath replied:
Added a comment.

mysterymath committed:
Address review comment; refactor

- Add block init comment
- Refactor out static inner_size/outer_size conversions
